controllers/application_controller.go

```diff
@@ -0,0 +1,247 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/retry"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appv1beta1 "sigs.k8s.io/application/api/v1beta1"
+)
+
+const (
+	loggerCtxKey = "logger"
+)
+
+// ApplicationReconciler reconciles a Application object
+type ApplicationReconciler struct {
+	client.Client
+	Mapper meta.RESTMapper
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups=app.k8s.io,resources=applications,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=app.k8s.io,resources=applications/status,verbs=get;update;patch
+
+func (r *ApplicationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	rootCtx := context.Background()
+	logger := r.Log.WithValues("application", req.NamespacedName)
+	ctx := context.WithValue(rootCtx, loggerCtxKey, logger)
+
+	var app appv1beta1.Application
+	err := r.Get(ctx, req.NamespacedName, &app)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	// Application is in the process of being deleted, so no need to do anything.
+	if app.DeletionTimestamp != nil {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus := r.getNewApplicationStatus(ctx, &app)
+
+	if equality.Semantic.DeepEqual(newApplicationStatus, &app.Status) {
+		return ctrl.Result{}, nil
+	}
+
+	newApplicationStatus.ObservedGeneration = app.Generation
+	if err = r.updateApplicationStatus(ctx, req.NamespacedName, newApplicationStatus); err != nil {
+		return ctrl.Result{Requeue: true}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *ApplicationReconciler) updateComponents(ctx context.Context, app *appv1beta1.Application) ([]*unstructured.Unstructured, error) {
+	resources, err := r.fetchComponentListResources(ctx, app.Spec.ComponentGroupKinds, app.Spec.Selector, app.Namespace)
+	if err != nil {
+		return resources, err
+	}
+
+	if app.Spec.AddOwnerRef {
+		ownerRef := metav1.NewControllerRef(app, appv1beta1.GroupVersion.WithKind("Application"))
+		*ownerRef.Controller = false
+		if err := r.setOwnerRefForResources(ctx, *ownerRef, resources); err != nil {
+			return resources, err
+		}
+	}
+	return resources, nil
+}
+
+func (r *ApplicationReconciler) getNewApplicationStatus(ctx context.Context, app *appv1beta1.Application) *appv1beta1.ApplicationStatus {
+
+	resources, err := r.updateComponents(ctx, app)
```

Comment:
I think it would be cleaner if we kept updating the owner refs separate from computing the status. The first includes updates to resources not managed by the application controller (so relatively risky operations), while the second is just computing their status. Currently updating the owner refs looks like a side effect of computing the new application status.